### PR TITLE
Allow custom collaborative paths in Vtxo tapscripts

### PR DESCRIPTION
**Pull request by louisinger**
_This PR adds the `ConditionMultisigClosure`, a new collaborative path letting to implement custom contracts_

### Context

The current implementation of the Vtxo script leaves allow only 2 types of collaborative path:
* `MultisigClosure` : `(A + ..... + S)`
* `CLTVMutisigClosure` : `(A + ..... + S) after t`

This scheme does not allow custom script to be used as collaborative path. 

### Conditional script

This PR introduce a new collaborative path pattern:

`<boolean_script> AND (A + .... + S)` 

Where `boolean_script` is any kind of bitcoin script with the following rules:
* returns 1 element on the stack.
* does not have any CHECKSIG-like opcodes.
* does not have any locktime opcodes `CSV` and `CLTV`.

Example of a valid conditional multisig closure:
```
OP_SHA256
OP_DATA_32
<hash>
OP_EQUAL
OP_VERIFY
<alice>
OP_CHECKSIGVERIFY
<server>
OP_CHECKSIG
```

To spend this collaborative path, the signer must provide, beside signatures, the witness arguments required to satisfy the `SHA256 <hash> EQUAL` script. The server will execute off-chain the script to validate the witness. 

@tiero @altafan please review

### CODE REVIEW

File: common/tree/script.go [hunk 1]
```diff
@@ -68,12 +69,21 @@ type CLTVMultisigClosure struct {
 	Locktime common.Locktime
 }
 
+// ConditionMultisigClosure is a closure that contains a condition and a
+// multisig closure. The condition is a boolean script that is executed with the
+// multisig witness.
+type ConditionMultisigClosure struct {
+	MultisigClosure
```
Comment (altafan) [inline common/tree/script.go:91]: shouldn't it be extension of `CLTVMultisigClosure`, which is itself an extension of `MultisigClosure`?

File: common/tree/script.go [hunk 2]
```diff
@@ -764,3 +774,149 @@ func (d *CLTVMultisigClosure) Decode(script []byte) (bool, error) {
...
+		witness[i], err = wire.ReadVarBytes(r, 0, txscript.MaxScriptSize, "witness")
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return witness, nil
+}
+func (f *ConditionMultisigClosure) WitnessSize() int {
+	return f.MultisigClosure.WitnessSize()
```
Comment (altafan) [inline common/tree/script.go]: this means we assume the condition not to affect the witness size?

File: common/tree/script_test.go [hunk 3]
```diff
@@ -624,3 +628,366 @@ func TestCLTVMultisigClosure(t *testing.T) {
...
+		},
+		{
+			name:        "invalid SHA256",
+			script:      append(append([]byte{txscript.OP_SHA256, txscript.OP_DATA_32}, hash1[:]...), txscript.OP_EQUAL),
+			witness:     wire.TxWitness{rand2},
+			returnValue: false,
+			expectErr:   false,
+		},
+		{
+			name: "valid coinflip different lengths",
```
Comment (altafan) [inline common/tree/script_test.go]: remove this test

File: common/tree/script_test.go [hunk 4]
```diff
@@ -624,3 +628,366 @@ func TestCLTVMultisigClosure(t *testing.T) {
...
+				txscript.OP_EQUAL,
+				txscript.OP_ENDIF,
+				txscript.OP_ENDIF,
+			}...),
+			witness:     wire.TxWitness{rand1, rand2},
+			returnValue: false,
+			expectErr:   false,
+		},
+		{
+			name: "valid coinflip same lengths",
```
Comment (altafan) [inline common/tree/script_test.go]: also this

File: common/tree/script_test.go [hunk 5]
```diff
@@ -624,3 +628,366 @@ func TestCLTVMultisigClosure(t *testing.T) {
...
+				txscript.OP_EQUAL,
+				txscript.OP_ENDIF,
+				txscript.OP_ENDIF,
+			}...),
+			witness:     wire.TxWitness{rand1, rand1},
+			returnValue: true,
+			expectErr:   false,
+		},
+		{
+			name: "invalid first element length",
```
Comment (altafan) [inline common/tree/script_test.go]: and also this one

File: common/tree/script_test.go [hunk 6]
```diff
@@ -624,3 +628,366 @@ func TestCLTVMultisigClosure(t *testing.T) {
...
+		decodedClosure := &tree.ConditionMultisigClosure{}
+		valid, err := decodedClosure.Decode(script)
+		require.NoError(t, err)
+		require.True(t, valid)
+		require.Equal(t, 1, len(decodedClosure.PubKeys))
+		require.True(t, bytes.Equal(conditionScript, decodedClosure.Condition))
+		require.True(t, bytes.Equal(schnorr.SerializePubKey(pubkey1), schnorr.SerializePubKey(decodedClosure.PubKeys[0])))
+	})
+
+	t.Run("valid condition with two key multisig", func(t *testing.T) {
```
Comment (altafan) [inline common/tree/script_test.go]:
```suggestion
	t.Run("valid condition with 2 of 2 multisig", func(t *testing.T) {
```

File: common/tree/script.go [hunk 7]
```diff
@@ -764,3 +774,149 @@ func (d *CLTVMultisigClosure) Decode(script []byte) (bool, error) {
...
+}
+
+func (f *ConditionMultisigClosure) Witness(controlBlock []byte, args map[string][]byte) (wire.TxWitness, error) {
+	script, err := f.Script()
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate script: %w", err)
+	}
+
+	// Read and execute condition witness
+	condWitness, err := ReadTxWitness(args["condition"])
```
Comment (altafan) [inline common/tree/script.go]: use constant

File: common/tree/script.go (hunk 1 above)
Comment (louisinger) [reply common/tree/script.go:91]: It implies several changes in the wallet and application server. As discussed offline, let's do it in a dedicated PR: https://github.com/ark-network/ark/issues/393

File: common/tree/script.go [hunk 8]
```diff
@@ -782,8 +781,47 @@ func (d *CLTVMultisigClosure) Decode(script []byte) (bool, error) {
...
+	// make sure the script doesn't contain any introspections opcodes (sig or locktime)
+	forbiddenOpcodes := []byte{
+		txscript.OP_CHECKMULTISIG,
+		txscript.OP_CHECKSIG,
+		txscript.OP_CHECKSIGVERIFY,
+		txscript.OP_CHECKSIGADD,
+		txscript.OP_CHECKMULTISIGVERIFY,
+		txscript.OP_CHECKLOCKTIMEVERIFY,
+		txscript.OP_CHECKSEQUENCEVERIFY,
+	}
```
Comment (altafan) [inline common/tree/script.go]: better to move this in a global unexported variable

**Review by altafan — APPROVED** [review]

**Pull request merged**